### PR TITLE
Give warning on setting alluxio.logs.dir in ALLUXIO_JAVA_OPTS

### PR DESCRIPTION
**Pull request by apc999**
### What changes are proposed in this pull request?

Before

```
$ export ALLUXIO_JAVA_OPTS="-Dalluxio.logs.dir=/tmp"
$ bin/alluxio getConf alluxio.logs.dir
/Users/binfan/projects/alluxio/alluxio/logs
```

After 

```
$ export ALLUXIO_JAVA_OPTS="-Dalluxio.logs.dir=/tmp"
$ bin/alluxio getConf alluxio.logs.dir
Warning: setting alluxio.logs.dir through ALLUXIO_JAVA_OPTS will be ignored. Use environment variable ALLUXIO_LOGS_DIR instead
/Users/binfan/projects/alluxio/alluxio/logs
```

### Why are the changes needed?

An issue reported in https://github.com/Alluxio/alluxio/pull/14837

### Does this PR introduce any user facing changes?

No

### CODE REVIEW

**Comment by Xenorith** [comment]
if the java opts for alluxio.logs.dir is set, then all these places will not be using the correct logs dir:
```
$ grep -r "ALLUXIO_LOGS_DIR" bin                                                                 [13:12:52]
bin/alluxio-workers.sh:mkdir -p "${ALLUXIO_LOGS_DIR}"
bin/alluxio-workers.sh:ALLUXIO_TASK_LOG="${ALLUXIO_LOGS_DIR}/task.log"
bin/alluxio-start.sh:  if [[ ! -d "${ALLUXIO_LOGS_DIR}" ]]; then
bin/alluxio-start.sh:    echo "ALLUXIO_LOGS_DIR: ${ALLUXIO_LOGS_DIR}"
bin/alluxio-start.sh:    mkdir -p ${ALLUXIO_LOGS_DIR}
bin/alluxio-start.sh:    echo "Starting job master @ $(hostname -f). Logging to ${ALLUXIO_LOGS_DIR}"
bin/alluxio-start.sh:    (nohup ${BIN}/launch-process job_master > ${ALLUXIO_LOGS_DIR}/job_master.out 2>&1) &
bin/alluxio-start.sh:  echo "Starting job worker @ $(hostname -f). Logging to ${ALLUXIO_LOGS_DIR}"
bin/alluxio-start.sh:  (nohup ${BIN}/launch-process job_worker > ${ALLUXIO_LOGS_DIR}/job_worker.out 2>&1) &
bin/alluxio-start.sh:    echo "Starting job worker #$((c+1)) @ $(hostname -f). Logging to ${ALLUXIO_LOGS_DIR}"
bin/alluxio-start.sh:    nohup ${BIN}/launch-process job_worker > ${ALLUXIO_LOGS_DIR}/job_worker.out 2>&1) &
bin/alluxio-start.sh:    nohup ${BIN}/launch-process logserver > ${ALLUXIO_LOGS_DIR}/logserver.out 2>&1) &
bin/alluxio-start.sh:    echo "Starting secondary master @ $(hostname -f). Logging to ${ALLUXIO_LOGS_DIR}"
bin/alluxio-start.sh:    (nohup ${BIN}/launch-process secondary_master > ${ALLUXIO_LOGS_DIR}/secondary_master.out 2>&1) &
bin/alluxio-start.sh:    echo "Starting master @ $(hostname -f). Logging to ${ALLUXIO_LOGS_DIR}"
bin/alluxio-start.sh:    (JOURNAL_BACKUP="${journal_backup}" nohup ${BIN}/launch-process master > ${ALLUXIO_LOGS_DIR}/master.out 2>&1) &
bin/alluxio-start.sh:    nohup ${BIN}/launch-process hub_agent > ${ALLUXIO_LOGS_DIR}/hub_agent.out 2>&1) &
bin/alluxio-start.sh:    nohup ${BIN}/launch-process hub_manager > ${ALLUXIO_LOGS_DIR}/hub_manager.out 2>&1) &
bin/alluxio-start.sh:  echo "Starting proxy @ $(hostname -f). Logging to ${ALLUXIO_LOGS_DIR}"
bin/alluxio-start.sh:  (nohup ${BIN}/launch-process proxy > ${ALLUXIO_LOGS_DIR}/proxy.out 2>&1) &
bin/alluxio-start.sh:  echo "Starting worker @ $(hostname -f). Logging to ${ALLUXIO_LOGS_DIR}"
bin/alluxio-start.sh:     nohup ${BIN}/launch-process worker > ${ALLUXIO_LOGS_DIR}/worker.out 2>&1 ) &
bin/alluxio-start.sh:    echo "Restarting worker @ $(hostname -f). Logging to ${ALLUXIO_LOGS_DIR}"
bin/alluxio-start.sh:    (nohup ${BIN}/launch-process worker > ${ALLUXIO_LOGS_DIR}/worker.out 2>&1) &
bin/alluxio-masters.sh:mkdir -p "${ALLUXIO_LOGS_DIR}"
bin/alluxio-masters.sh:ALLUXIO_TASK_LOG="${ALLUXIO_LOGS_DIR}/task.log"
bin/alluxio-monitor.sh:  ALLUXIO_TASK_LOG="${ALLUXIO_LOGS_DIR}/task.log"
bin/alluxio-monitor.sh:  local node_logs=( "${ALLUXIO_LOGS_DIR}/${node_type}.log" "${ALLUXIO_LOGS_DIR}/${node_type}.out" )
```
so each of these scripts should evaluate the configured value of `alluxio.logs.dir` wherever `ALLUXIO_LOGS_DIR` is read as a value

**Comment by apc999** [comment]
I could manually read value of `alluxio.logs.dir` to assigne to `ALLUXIO_LOGS_DIR` in beginning of each script, I feel it is error prone and hacky (what if some one adding more script without knowing this convention?)

It looks to me env var `ALLUXIO_LOGS_DIR` shall be the only way to set log dir (similar to `ALLUXIO_HOME` and `ALLUXIO_CONF_DIR`).  They are anyway used and treated in a very different way from other conf properties.
To keep it consistent, I would simply hide `alluxio.logs.dir` and also `alluxio.home` from the public property list from users, as we are not expecting users to change them at all through property.

@Xenorith @yuzhu what's your thought?

**Comment by Xenorith** [comment]
this is the crux of the discussion in the prev PR and probably why it got stalled.

if we only allow the env var as the mechanism to set, we should somehow warn or disallow users who are trying to set the logs dir through the java opts. this is not a random corner case because it is definitely happening; more than one reporter have tried to use the java opts to set and wondering why it is not being respected. at the minimum, we have to do regex parsing in the java opts strings (and remember, there are multiple java opts for the process specific ones) to know whether or not to trigger a deprecation warning/failure. some of this work was done in the previous PR and both @jiacheliu3 and i noted how complex this logic is.

so seems like either choice, there is going to be yucky code to add, but which one is more yucky?

**Comment by Xenorith** [comment]
> Agree. I will
> 
> 1. warn if anyone is setting this property through java opts

suggest looking at this change as a starting point: https://github.com/Alluxio/alluxio/pull/14837/files#diff-9dd02c390f78580fa089e750abbed7dda6962da737f8fb6941e8800df3481aa0

> 2. hide this property from docs (I could not fully remove them as they are used in internal tests) --- so there will no way for users to know the existence of "alluxio.logs.dir" unless they read the src code.

is there also a way to hide it from being shown in the generated table: https://docs.alluxio.io/os/user/stable/en/reference/Properties-List.html

> 3. clean up docs
> 
> Going this way is fine -- at least no ugly code added

i guess this is ugly but not production effective. this also needs to be clearly stated in the next release notes

**Comment by apc999** [comment]
> > Agree. I will
> > 
> > 1. warn if anyone is setting this property through java opts
> 
> suggest looking at this change as a starting point: https://github.com/Alluxio/alluxio/pull/14837/files#diff-9dd02c390f78580fa089e750abbed7dda6962da737f8fb6941e8800df3481aa0
Ok
> 
> > 2. hide this property from docs (I could not fully remove them as they are used in internal tests) --- so there will no way for users to know the existence of "alluxio.logs.dir" unless they read the src code.
> 
> is there also a way to hide it from being shown in the generated table: https://docs.alluxio.io/os/user/stable/en/reference/Properties-List.html

yes  that's exactly what I mean -- hide it from this page

> 
> > 3. clean up docs
> > 
> > Going this way is fine -- at least no ugly code added
> 
> i guess this is ugly but not production effective. this also needs to be clearly stated in the next release notes
SG

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 1]
```diff
@@ -306,6 +306,7 @@ public String toString() {
           .setDefaultValue(String.format("${%s}/conf", Name.HOME))
           .setDescription("The directory containing files used to configure Alluxio.")
           .setIgnoredSiteProperty(true)
+          .setIsHidden(true)
```
Comment (Xenorith) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:309]: for these 3 unique properties that should be set through env, can the env var be mentioned somewhere in the property definition, ex. in its description?
```
.setDescription("The directory containing files used to configure Alluxio. This should only be set via the environment variable ALLUXIO_CONF_DIR when invoking startup scripts and never be set as a java opt or in alluxio-site.properties.")
```

File: core/common/src/main/java/alluxio/conf/PropertyKey.java [hunk 2]
```diff
@@ -376,12 +377,14 @@ public String toString() {
...
           .setScope(Scope.ALL)
           .build();
+  // Used in alluxio-config.sh and conf/log4j.properties
   public static final PropertyKey LOGS_DIR =
       new Builder(Name.LOGS_DIR)
           .setDefaultValue(String.format("${%s}/logs", Name.WORK_DIR))
           .setDescription("The path under Alluxio home directory to store log files. It has a "
               + "corresponding environment variable $ALLUXIO_LOGS_DIR.")
           .setIgnoredSiteProperty(true)
+          .setIsHidden(true)
```
Comment (Xenorith) [inline core/common/src/main/java/alluxio/conf/PropertyKey.java:387]: at least the two log properties mention the corresponding env var, but i think it's still better to include the full disclaimer.

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 1 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:309]: done

File: core/common/src/main/java/alluxio/conf/PropertyKey.java (hunk 2 above)
Comment (apc999) [reply core/common/src/main/java/alluxio/conf/PropertyKey.java:387]: done

File: libexec/alluxio-config.sh [hunk 3]
```diff
@@ -67,6 +67,18 @@ if [[ -n "${ALLUXIO_HOME}" ]]; then
   ALLUXIO_JAVA_OPTS+=" -Dalluxio.home=${ALLUXIO_HOME}"
 fi
 
+if [[ ${ALLUXIO_JAVA_OPTS} == *alluxio.conf.dir* ]]; then
```
[the remark this answers is not in the document]
Comment (apc999) [reply libexec/alluxio-config.sh:70]: done

**Comment by apc999** [comment]
PTAL @Xenorith @yuzhu

**Review by Xenorith — APPROVED** [review]

**Comment by apc999** [comment]
```
Error: 6.598 [ERROR] Errors: 
Error: 6.598 [ERROR]   LocationCommandIntegrationTest>AbstractFileSystemShellTest.beforeClass:122 » Runtime
```

**Review by yuzhu — APPROVED** [review]
I agree this is the cleanest solution. LGTM

**Comment by flaming-archer** [comment]
LGTM. from  #14837

**Pull request merged**